src/main/scala/amba/apb/ToTL.scala

```diff
@@ -0,0 +1,83 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.amba.apb
+
+import Chisel._
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util.MaskGen
+
+case class APBToTLNode()(implicit valName: ValName) extends MixedAdapterNode(APBImp, TLImp)(
+  dFn = { case APBMasterPortParameters(masters) =>
+    TLClientPortParameters(clients = masters.map { m =>
+      TLClientParameters(name = m.name, nodePath = m.nodePath)
+    })
+  },
+  uFn = { mp => APBSlavePortParameters(
+    slaves = mp.managers.map { m =>
+      APBSlaveParameters(
+        address       = m.address,
+        resources     = m.resources,
+        regionType    = m.regionType,
+        executable    = m.executable,
+        nodePath      = m.nodePath,
+        //TODO why and how should I do this adjust for APB?
+        // Also, shouldn't I use supportsPutPartial because of PSTRB?
+        supportsWrite = m.supportsPutFull,//adjust(m.supportsPutFull),
+        supportsRead  = m.supportsGet)},//adjust(m.supportsGet))},
+    beatBytes = mp.beatBytes)
+  }) {
+  //TODO: Shouldn't there be a requirement that TL side is wider than APB side?
+}
+
+class APBToTL()(implicit p: Parameters) extends LazyModule
+{
+  val node = APBToTLNode()
+
+  lazy val module = new LazyModuleImp(this) {
+    (node.in zip node.out) foreach { case ((in, edgeIn), (out, edgeOut)) =>
+      val beatBytes = edgeOut.manager.beatBytes
+
+      //-----------------------------------------------
+      // Flow Control
+      out.a.valid := in.psel && in.penable
+      in.pready := in.psel & in.penable & out.d.valid
+      out.d.ready := in.psel && in.penable // No backpressure to APB Master
```

Comment:
This is wrong. You can't just arbitrarily back-pressure D. I think what you actually want is:

 in.pready := out.d.valid
 out.d.ready := true.B
 assert (!out.d.valid || (in.penable && in.psel))

The APB spec guarantees that penable and psel stay high until pready is asserted by the slave.

Reply:
There's another problem, which is that TileLink might respond on the same cycle as the request. This is too early for APB to handle. However, the obvious fix of back-pressuring d.ready when !sent would violate the TileLink forward progress rules. The only way to be fully compliant would be a data-bus wide Queue to capture the D response. However, this corner case is probably not relevant most of the time, so I'd instead just add a requirement on diplomacy:

require (edgeOut.manager.minLatency >= 1)

Reply:
It's not too early for APB to handle if you only look at PENABLE, right?  I thought that the first cycle (PSEL = 1) is sort of the "get ready" and "PENABLE = 1" means "go". And if that PREADY is high right away, then the read data is also ready.

In other words, PSEL is pretty much pointless for our conversion.

Though I could be reading the spec wrong.

Reply:
No. PSEL is "go" and PENABLE is "now it's your turn".

Reply:
The old APB and AHB specs are from an era when you had this whole bus-turn-around nonsense. You could imagine PENABLE to be like the direction of the data bus. PENABLE=0 means master controls it and PENABLE=1 means the slave controls it.

Reply:
I somewhat changed your suggested assertion above. I want the transaction to go out on PSEL=1, but not come back until PENABLE=1, so this is the assertion I wrote:

`      assert (!(in.psel && !in.penable && out.d.valid))
`
Make sense?

Reply:
By de morgan's laws, here is my assertion in your form: (!(out.d.valid && (!in.penable || !in.psel))
You can see it already covers your scenario, but additionally requires that psel is 1 when dvalid is 1.